site/guides/effective-go-krm-function.md

```diff
@@ -0,0 +1,212 @@
+# Effective GO KRM functions
+
+This guide gives tips to effectively write a KRM function. 
+
+This guide is for advanced kpt function users who find the [catalog.kpt.dev] cannot fulfil their needs 
+and want to design their own KRM functions. 
+
+Suggest reading [Developing in Go] first
+
+## Prerequisites
+
+- [Install kpt]
+- [Install Docker]
+
+## Setup 
+
+<!--- TODO: Use scaffolding to generate the get-started package --->
+
+We start from a "get-started-runner" package which contains a `main.go` file with some scaffolding code.
+
+
+```shell
+# Set your KRM function name.
+export FUNCTION_NAME=<YOUR FUNCTION NAME>
+export GOPATH=$(go env GOPATH)
+export FUNCTION_PATH=github.com/<YOUR USERNAME>
+
+# Create and direct to your Go working directory
+mkdir -p $GOPATH/src/${FUNCTION_PATH} && cd $GOPATH/src/${FUNCTION_PATH}
+
+# Get the "get-started" package.
+kpt pkg get https://github.com/GoogleContainerTools/kpt-functions-sdk.git/go/get-started-runner@master ${FUNCTION_NAME}
+
+cd ${FUNCTION_NAME}
+
+# Initialize Go module and install the kpt KRM function SDK.
+go mod init && go mod tidy -compat=1.17
+```
+
+## Write your KRM function code in Go 
+
+In the main.go, you should have
+```go
+var _ fn.Runner = &FunctionX{}
+
+type FunctionX struct {
+    // TODO: Modify with your expected function config.
+    FnConfigBool bool
+    FnConfigInt  int
+    FnConfigFoo string
+}
+
+func (r *FunctionX) Run(ctx *fn.Context, functionConfig *fn.KubeObject, items fn.KubeObjects) {
+      // TODO: Add your KRM resource mutate or validate logic.
+}
+
+func main() {
+    if err := fn.AsMain(&FunctionX{}); err != nil {
+        os.Exit(1)
+    }
+}
+```
+`FunctionX` implements the [`Runner`] interface that can process the input KRM resources as [`fn.ResourceList`], it initializes `fn.KubeObject` to hold the KRM resources,
+so that you can use [`fn.KubeObject` and `fn.SubObject`] methods directly. After `Run`, it will convert the modified `fn.KubeObjects` to KRM resources.
+
+### Define configures
+
+If you need to use configurable variables, you can define them as `FunctionX` fields.
+Otherwise you can skip this step and move to next. 
+```go 
+type FunctionX struct {
+  FnConfigBool bool
+  FnConfigInt  int
+  FnConfigFoo string
+}
+```
+
+For example, define a `SetImage` and add two variables to compare-and-swap the image value will be like:
+```go
+type SetImage struct {
+  OldImage string // Existing image
```

Comment:
This is good - the example makes it much clearer.  We might consider just walking through SetImage, instead of FunctionX.  Start with no parameters, and then say "we want to change an image from an old image value to a new image value, so we need parameters so the user can specify the old image and the new image" or something like that